patroni/postgresql/validator.py

```diff
@@ -128,456 +142,371 @@ def transform(self, name: str, value: Optional[Any]) -> Optional[Any]:
+        :raises :class:`ValidatorFactoryInvalidSpec`: if *validator* contains an invalid set of attributes for the
+            given ``type``.
+
+        :Example:
+
+            If a given validator was defined as follows in the YAML file:
+
+            ```yaml
+            - type: String
+            version_from: 90300
+            version_till: null
+            ```
+
+            Then this method would receive *validator* as:
+
+            ```python
+            {
+                'type': 'String',
+                'version_from': 90300,
+                'version_till': None
+            }
+            ```
+
+            And this method would return a :class:`String`:
+
+            ```python
+            String(90300, None)
+            ```
+        """
+        validator = deepcopy(validator)
+        try:
+            type_ = validator.pop('type')
+        except KeyError as exc:
+            raise ValidatorFactoryNoType('Validator contains no type.') from exc
+
+        if type_ not in cls.TYPES:
+            raise ValidatorFactoryInvalidType(f'Unexpected validator type: `{type_}`.')
+
+        for key, value in validator.items():
+            # :func:`_transform_parameter_value` expects :class:`tuple` instead of :class:`list`
+            if isinstance(value, list):
+                tmp_value: List[Any] = value
+                validator[key] = tuple(tmp_value)
+
+        try:
+            return cls.TYPES[type_](**validator)
+        except Exception as exc:
+            raise ValidatorFactoryInvalidSpec(
+                f'Failed to parse `{type_}` validator (`{validator}`): `{str(exc)}`.') from exc
+
+
+def _load_postgres_guc_validators(section: CaseInsensitiveDict, config: Dict[str, Any], parameter: str) \
+        -> Iterator[Union[ValidatorFactoryNoType, ValidatorFactoryInvalidType, ValidatorFactoryInvalidSpec]]:
+    """Load *parameter* validators from *config* into *section*.
+
+    Loop over all validators of *parameter* and load each of them into *section*.
+
+    :param section: reference to either ``parameters`` or ``recovery_parameters`` variable of this module.
+    :param config: Python object corresponding to an YAML file, with values of either ``parameters`` or
+        ``recovery_parameters`` key, corresponding to given *section*.
+    :param parameter: name of the parameter found under *config* which validators should be parsed and loaded into
+        *section*.
+
+    :rtype: yields any exception that is faced while parsing a validator spec into a Patroni validator object.
+    """
+    for validator_spec in config.get(parameter, []):
+        try:
+            validator = ValidatorFactory(validator_spec)
+        except (ValidatorFactoryNoType, ValidatorFactoryInvalidType, ValidatorFactoryInvalidSpec) as exc:
+            yield exc
+            continue
+
+        section[parameter] = section.get(parameter, ()) + (validator,)
```

Comment:
Hmm.
Lets say we have two yaml files.
The default one, 0_postgres.yml and 1_vendor.yml.
The first one defines:
```yaml
  wal_compression:
  - type: Bool
    version_from: 90500
    version_till: 150000
  - type: EnumBool
    version_from: 150000
    possible_values:
    - pglz
    - lz4
    - zstd
```

And the second one:
```yaml
  wal_compression:
  - type: Bool
    version_from: 90500
    version_till: 150000
  - type: EnumBool
    version_from: 150000
    possible_values:
    - brotli
    - pglz
    - lz4
    - zstd
```

With this approach we will have a tuple with 4 objects: `(Bool(), EnumBool(), Bool(), EnumBool())`, what is IMO wrong.
If the same parameter is defined in a different file, it should be completely overwritten.

IMO, this function should simply receive `config[parameter]` as an argument and return a tuple.
All exceptions that are currently yielded from here should be just converted to warnings.

What do you think @barthisrael?